### PR TITLE
ui: decred privacy ui

### PR DESCRIPTION
**Pull request by buck54321**
Implement core, webserver and ui components of new Decred mixing functionality. Refactor some wallet code to accommodate new simnet (non)mixer, which just sends funds from unmixed to mixed every block. Add `ExtensionModeConfig` fields for disabling staking and mixing through the gui.

image

image

image

image

image

### CODE REVIEW

**Comment by peterzen** [comment]
Getting this `api error` message:
output

**Comment by peterzen** [comment]
The layout of the accounts make it look like Unmixed is the sum of the amounts above it:

Screenshot 2023-11-20 06:34:34

**Comment by peterzen** [comment]
*Staking* and *Enable privacy* look very similar, but Staking is a heading and *Enable privacy* is actually a button (this is visually not obvious).  It's not clear from the user's perspective what the latter does until you click it and it creates the Unmixed account.
image

Is there a reason for having this extra step, i.e. not displaying the toggle right away and doing the account creation when that is enabled?

image

**Comment by peterzen** [comment]
The appearance and the layout the buttons on this dialog in inconsistent with those on the other dialogs (the buttons appear more as footers visually)

Screenshot 2023-11-20 06:35:31

**Comment by buck54321** [comment]
> The layout of the accounts make it look like Unmixed is the sum of the amounts above it:
> 
> Screenshot 2023-11-20 06:34:34

I'm going to tackle a wallets page refactor this week with the primary goal of getting the balances section to make sense.

**Comment by buck54321** [comment]
> Getting this `api error` message: output

Hmm. I thought I fixed that. I'll get it figured out.

File: client/asset/dcr/spv_mixing.go [hunk 1]
```diff
@@ -13,124 +10,29 @@ const (
...
-		conn, err := new(net.Dialer).DialContext(ctx, network, addr)
-		if err != nil {
-			return nil, err
-		}
-		conn = tls.Client(conn, csppTLSConfig)
-		return conn, nil
+	// Don't perform any actions while transactions are not synced
+	// through the tip block.
+	if !w.spv.Synced() {
+		w.log.Debugf("Skipping autobuyer actions: transactions are not synced")
```
Comment (JoeGruffins) [inline client/asset/dcr/spv_mixing.go:31]: Is `autobuyer` correct? Should say mixer? the autobuyer is for automatic ticket purchases correct?

File: client/core/core.go [hunk 2]
```diff
@@ -10878,3 +10879,77 @@ func (c *Core) TicketPage(assetID uint32, scanStart int32, n, skipN int) ([]*ass
...
+		if err != nil {
+			return fmt.Errorf("unlock error: %w", err)
+		}
+	}
+	return mw.StartFundsMixer(c.ctx)
+}
+
+// StopFundsMixer stops the funds mixer. This will error if the wallet does
+// not allow starting or stopping the mixer or if the mixer was not already
+// running.
```
Comment (JoeGruffins) [inline client/core/core.go:10937]: Does not currently error for the reasons stated.

File: client/webserver/locales/en-us.go [hunk 3]
```diff
@@ -497,4 +497,19 @@ var EnUS = map[string]string{
...
+	"Disable Privacy":             "Disable Privacy",
+	"Configure Privacy":           "Configure Privacy",
+	"Privacy active":              "Privacy active",
+	"Privacy off":                 "Privacy off",
+	"staking_disabled":            `Ticket purchasing disabled by <span id="extensionModeAppName"></span>.`,
+	"loading privacy status":      "loading privacy status",
+	"mixing_pw_prompt":            "Enter your password to unlock your wallet turn on privacy",
+	"cspp_addr":                   "CSPP Server Address",
+	"cspp_how":                    "CoinShuffle++ creates outputs that cannot be definitively linked to previous on-chain activity. While output amounts are still visible in this form of privacy, the destruction of linkability means your on-chain history can no longer be inferred from publicly available data.",
+	"privacy_intro":               "When privacy is enabled, all of your funds will be sent through an address history obfuscation service using a protocol called CoinShuffle++.",
```
Comment (martonp) [inline client/webserver/locales/en-us.go:512]: Would be helpful to know that each time you receive funds, you'll have to wait for an additional block to get mined (or maybe more?) before you can use your funds.

Also, a comment on the difference between turning off privacy and disabling privacy would be helpful.

File: client/asset/interface.go [hunk 4]
```diff
@@ -704,27 +704,34 @@ type FeeRater interface {
...
+	// // MixedBalance is the amount of funds that have been successfully mixed and
+	// // may be withdrawn or used to fund trades.
+	// MixedBalance uint64 `json:"mixedBalance"`
+	// // UnmixedBalance is the amount of funds that are available and ready for
+	// // mixing. If the wallet is not configured for mixing, this balance may be
+	// // withdrawn or used to fund trades.
+	// UnmixedBalance uint64 `json:"unmixedBalance"`
+	// // UnmixedBalanceThreshold is the minimum amount of unmixed funds that must
+	// // be in the wallet for mixing to happen.
+	// UnmixedBalanceThreshold uint64 `json:"unmixedBalanceThreshold"`
```
Comment (martonp) [inline client/asset/interface.go:726]: Can remove.

**Review by martonp — COMMENTED** [review]
Looking good and working well.

**Review by martonp — APPROVED** [review]

**Comment by buck54321** [comment]
> Seeing `wallet.MixOutput(c8333c4d804cc4c1f6ae385c8e21aded1ed853f10640a65045582f6c8ccecda1:0): dial tcp 149.28.110.221:15760: connect: connection refused` but I _think_ the testnet mixer is running because I am using it on some other machines. Also I am able to ping it.
> 
> edit: Actually it seems the testnet server is just down. The other machines are not able to use it either.
> 
> It also seems like my funds in my testnet dexc wallet are now stuck in the `unmixed` account? Should they be moved after turning off mixing?

Still seeing this today. Will hold off on merge until I can verify on testnet. Otherwise ready.

Funds are moved when privacy is disabled, not when it is just turned off.

**Pull request merged**